index.js

```diff
@@ -1,40 +1,60 @@
-'use strict';
 const local = require('./inject.js');
 const remote = require('./request.js');
 const querystring = require('querystring');
+const Boom = require('boom');
 
-const register = async function(server, pluginOptions) {
-  const callIt = async (method, url, options) => {
-    Object.assign(options, pluginOptions);
-    // construct url from any relevant options:
-    if (options.query) {
-      url += `?${querystring.stringify(options.query)}`;
-    }
-    if (url[0] === '/') {
-      if (pluginOptions.localPrefix) {
-        url = `${pluginOptions.localPrefix}${url}`;
+const defaults = {
+  maxRetries: 0
+};
+
+const register = function(server, pluginOptions) {
+  const callIt = async (method, url, options, count = 0) => {
+    try {
+      let response;
+
+      Object.assign(options, defaults, pluginOptions);
+
+      // construct url from any relevant options:
+      if (options.query) {
+        url += `?${querystring.stringify(options.query)}`;
+      }
+      if (url[0] === '/') {
+        if (pluginOptions.localPrefix) {
+          url = `${pluginOptions.localPrefix}${url}`;
+        }
+
+        response = await local(server, method, url, options);
+      } else {
+        response = await remote(method, url, options);
+      }
+
+      if (Boom.isBoom(response) || response instanceof Error) {
+        throw Boom.boomify(response);
       }
-      return await local(server, method, url, options);
+
+      return response;
+    } catch (e) {
+      if (count < options.maxRetries) {
+        server.log(['hapi-req', 'info'], `Retry #${count + 1}: ${method} ${url}`);
+        return callIt(method, url, options, count + 1);
+      }
+
+      if (options.maxRetries) {
+        server.log(['hapi-req', 'info'], `Max retries: ${method} ${url}`);
+      }
+
+      return e;
```

Comment:
Should it throw this?

Reply:
I think returning this one is better since we can do a if response.isBoom instead of having to put in another try/catch on the requesting side.

Reply:
but the problem is that we are assuming the payload is coming back, not a response.  I think we decided that libraries should always throw and the main app will decide if it wants to do anything with that.